src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php

```diff
@@ -275,6 +275,17 @@ protected function insert_attachment( $request ) {
 		$attachment->post_mime_type = $type;
 		$attachment->guid           = $url;
 
+		// If the title was not set, use the original filename.
+		if ( empty( $attachment->post_title ) && ! empty( $files['file']['name'] ) ) {
+			// Remove the file extension (after the last `.`)
+			$tmp_title = substr( $files['file']['name'], 0, strrpos( $files['file']['name'], '.' ) );
+
+			if ( ! empty( $tmp_title ) ) {
+				$attachment->post_title = $tmp_title;
+			}
+		}
+
+		// Fall back to the original approach.
```

Comment:
Rather than run `empty( $attachment->post_title )` twice when it's not empty, what about:

```php
if ( empty( $attachment->post_title ) ) {
    if ( ! empty( $files['file']['name'] ) ) {
        $attachment->post_title = preg_replace( '/\.[^.]+$/', '', $files['file']['name'] );
    } else {
        $attachment->post_title = preg_replace( '/\.[^.]+$/', '', wp_basename( $file ) );
    }
}
```

One gotcha is a dot file, but this isn't handled in the original code either (but possibly by `wp_insert_attachment()` later.